geomesa-memory/src/main/scala/org/locationtech/geomesa/memory/cqengine/GeoCQEngine.scala

```diff
@@ -0,0 +1,188 @@
+  // TODO: Add logic to allow for the geo-index to be disabled?  (Low priority)
+  addGeoIndex(sft.getGeometryDescriptor)
+
+  if (enableFidIndex) addFidIndex()
+
+  // Add other indexes
+  sft.getAttributeDescriptors.foreach {addIndex(_)}
+
+  def remove(sf: SimpleFeature): Boolean = {
+    cqcache.remove(sf)
+  }
+
+  def add(sf: SimpleFeature): Boolean = {
+    cqcache.add(sf)
+  }
+
+  def addAll(sfs: util.Collection[SimpleFeature]): Boolean = {
+    cqcache.addAll(sfs)
+  }
+
+  def clear(): Unit = {
+    cqcache.clear()
+  }
+
+  def getById(id: String): Option[SimpleFeature] = {
+    // if this gets used, set enableFidIndex=true
+    cqcache.retrieve(new Equal(SFTAttributes.fidAttribute, id)).headOption
+  }
+
+  def update(sf: SimpleFeature): Boolean = {
+    getById(sf.getID) match {
+      case Some(oldsf) => cqcache.remove(oldsf)
+      case None =>
+    }
+    cqcache.add(sf)
+  }
+
+  // NB: We expect that FID filters have been handled previously
+  def getReaderForFilter(filter: Filter): FR =
+    filter match {
+      case f: IncludeFilter => include(f)
+      case f => queryCQ(f, dedup = true)
+    }
+
+  def queryCQ(f: Filter, dedup: Boolean = true): FR = {
+    val visitor = new CQEngineQueryVisitor(sft)
+
+    val query: Query[SimpleFeature] = f.accept(visitor, null) match {
+      case q: Query[SimpleFeature] => q
+      case _ => throw new Exception(s"Filter visitor didn't recognize filter: $f.")
+    }
+    if (dedup) {
+      val dedupOpt = QueryFactory.deduplicate(DeduplicationStrategy.LOGICAL_ELIMINATION)
+      val queryOptions = QueryFactory.queryOptions(dedupOpt)
+      new DFR(sft, new DFI(cqcache.retrieve(query, queryOptions).iterator()))
+    }
+    else
+      new DFR(sft, new DFI(cqcache.retrieve(query).iterator()))
+  }
+
+  def include(i: IncludeFilter): FR = {
+    logger.warn("Running Filter.INCLUDE")
+    new DFR(sft, new DFI(cqcache.retrieve(new All(classOf[SimpleFeature])).iterator()))
+  }
+
+  private def addIndex(ad: AttributeDescriptor): Unit = {
+    CQIndexingOptions.getCQIndexType(ad) match {
+      case CQIndexType.DEFAULT =>
+        ad.getType.getBinding match {
+          // Comparable fields should have a Navigable Index
+          case c if
+          classOf[java.lang.Integer].isAssignableFrom(c) ||
+            classOf[java.lang.Long].isAssignableFrom(c) ||
+            classOf[java.lang.Float].isAssignableFrom(c) ||
+            classOf[java.lang.Double].isAssignableFrom(c) ||
+            classOf[java.util.Date].isAssignableFrom(c) => addNavigableIndex(ad)
+          case c if classOf[java.lang.String].isAssignableFrom(c) => addRadixIndex(ad)
+          case c if classOf[Geometry].isAssignableFrom(c) => addGeoIndex(ad)
+          case c if classOf[UUID].isAssignableFrom(c) => addUniqueIndex(ad)
+          // TODO: Decide how boolean fields should be indexed
+          case c if classOf[java.lang.Boolean].isAssignableFrom(c) => addHashIndex(ad)
```

Comment:
Should add a default case to be safe. Also, I think you're missing some cases like for List or Map: https://github.com/locationtech/geomesa/blob/master/geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/SimpleFeatureTypes.scala#L530